<!DOCTYPE html>
<html lang="en">
  <head>
<meta charset="utf-8" />
<meta
  name="viewport"
  content="width=device-width, initial-scale=1.0, maximum-scale=1.0"
/>
<meta
  name="description"
  content="Design and implement your programming language and software analysis tools with mathematical rigor."
/>
<meta name="keywords" content="runtime, verification, rv, k" />
<meta name="author" content="IELE Semantics | Runtime Verification Inc" />
<meta name="robots" content="index, follow" />

<!--favicon icon-->
<link rel="icon" type="image/png" href="../../../../../assets/img/favicon.ico" />

<title>IELE Semantics | Runtime Verification Inc</title>

<!-- <base href="/k/" /> -->

<!--web fonts-->
<link
  href="https://fonts.googleapis.com/css?family=Nunito:300,400,600,700,800"
  rel="stylesheet"
/>
<link
  href="https://fonts.googleapis.com/css?family=Lora:400i"
  rel="stylesheet"
/>

<link
  href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.14.0/css/all.min.css"
  rel="stylesheet"
/>

<link href="../../../../../assets/css/index.css" rel="stylesheet" />

<!--[if (gt IE 9) |!(IE)]><!-->
<!--<link rel="stylesheet" href="/assets/vendor/custom-nav/css/effects/fade-menu.css"/>-->
<!-- <link rel="stylesheet" href="../../../../../assets/k/vl-nav/css/effects/slide-menu.css" /> -->
<!--<![endif]-->

  </head>

  <body>
<header
  class="navbar navbar-expand navbar-dark flex-column flex-md-row bd-navbar"
>
  <a class="logo-link" href="../../../../../index.html">
    <img
      class="logo-dark"
      srcset="../../../../../assets/img/k-logo.png"
      alt="K"
      style="height: 48px"
    />
    Semantic Framework
  </a>
  <ul class="navbar-nav ml-md-auto">
    <li class="nav-item">
      <a
        class="nav-link pl-2 pr-1 mx-1 py-3 my-n2"
        href="https://github.com/runtimeverification/iele-semantics"
        target="_blank"
        rel="noopener"
        aria-label="GitHub"
      >
        <i class="fab fa-github"></i>
      </a>
    </li>
  </ul>
  <a
    class="btn btn-rv-blue d-none d-lg-inline-block mb-3 mb-md-0 ml-md-3"
    href="../../../../../downloads"
    >Download</a
  >
</header>


    <div class="container-fluid">
      <div class="row flex-xl-nowrap">
<div class="col-12 col-md-3 col-xl-2 bd-sidebar mb-3">
  <form
    role="search"
    class="bd-search d-flex align-items-center justify-content-between"
  >
    <input
      type="search"
      class="form-control"
      placeholder="Search..."
      aria-label="Search for..."
      autocomplete="false"
      id="search-box"
    />
    <button
      class="btn bd-search-docs-toggle d-md-none p-0 ml-3 collapsed"
      type="button"
      aria-label="Toggle docs navigation"
      style="font-size: 1.4rem"
    >
      <i class="fas fa-bars"></i>
    </button>
  </form>
  <nav class="collapse bd-links" aria-label="Main navigation">
    <div class="bd-toc-item">
      <a class="bd-toc-link" href="../../../../../">Homepage</a>
      <a class="bd-toc-link" href="../../../../../INSTALL">Install</a>
      <a class="bd-toc-link" href="../../../../../Design">Design</a>
    </div>
  </nav>
</div>

        <main
          class="col-12 col-md-9 col-xl-8 py-md-3 pl-md-5 bd-content"
          role="main"
        >
          <div class="markdown-preview"><html><head></head><body><h1>Error handling</h1>
<p>Error handling represents one of the most important considerations when
implementing a Node.js native add-on. When an error occurs in your C++ code you
have to handle and dispatch it correctly. <strong>node-addon-api</strong> uses return values and
JavaScript exceptions for error handling. You can choose return values or
exception handling based on the mechanism that works best for your add-on.</p>
<p>The <code>Napi::Error</code> is a persistent reference (for more info see: <a href="object_reference/"><code>Napi::ObjectReference</code></a>)
to a JavaScript error object. Use of this class depends on whether C++
exceptions are enabled at compile time.</p>
<p>If C++ exceptions are enabled (for more info see: <a href="setup/">Setup</a>), then the
<code>Napi::Error</code> class extends <code>std::exception</code> and enables integrated
error-handling for C++ exceptions and JavaScript exceptions.</p>
<p>The following sections explain the approach for each case:</p>
<ul>
<li><a href="#exceptions">Handling Errors With C++ Exceptions</a></li>
<li><a href="#noexceptions">Handling Errors Without C++ Exceptions</a></li>
</ul>
<p><a name="exceptions"></a></p>
<p>In most cases when an error occurs, the addon should do whatever clean is possible
and then return to JavaScript so that they error can be propagated.  In less frequent
cases the addon may be able to recover from the error, clear the error and then
continue.</p>
<h2>Handling Errors With C++ Exceptions</h2>
<p>When C++ exceptions are enabled try/catch can be used to catch exceptions thrown
from calls to JavaScript and then they can either be handled or rethrown before
returning from a native method.</p>
<p>If a node-addon-api call fails without executing any JavaScript code (for example due to
an invalid argument), then node-addon-api automatically converts and throws
the error as a C++ exception of type <code>Napi::Error</code>.</p>
<p>If a JavaScript function called by C++ code via node-addon-api throws a JavaScript
exception, then node-addon-api automatically converts and throws it as a C++
exception of type <code>Napi:Error</code> on return from the JavaScript code to the native
method.</p>
<p>If a C++ exception of type <code>Napi::Error</code> escapes from a N-API C++ callback, then
the N-API wrapper automatically converts and throws it as a JavaScript exception.</p>
<p>On return from a native method, node-addon-api will automatically convert a pending C++
exception to a JavaScript exception.</p>
<p>When C++ exceptions are enabled try/catch can be used to catch exceptions thrown
from calls to JavaScript and then they can either be handled or rethrown before
returning from a native method.</p>
<h2>Examples with C++ exceptions enabled</h2>
<h3>Throwing a C++ exception</h3>
<pre class="hljs"><code>Env env = ...
<span class="hljs-keyword">throw</span> Napi::Error::New(env, <span class="hljs-string">&quot;Example exception&quot;</span>);
<span class="hljs-comment">// other C++ statements</span>
<span class="hljs-comment">// ...</span>
</code></pre>
<p>The statements following the throw statement will not be executed. The exception
will bubble up as a C++ exception of type <code>Napi::Error</code>, until it is either caught
while still in C++, or else automatically propagated as a JavaScript exception
when returning to JavaScript.</p>
<h3>Propagating a N-API C++ exception</h3>
<pre class="hljs"><code>Napi::Function jsFunctionThatThrows = someObj.As&lt;Napi::Function&gt;();
Napi::Value result = jsFunctionThatThrows({ arg1, arg2 });
<span class="hljs-comment">// other C++ statements</span>
<span class="hljs-comment">// ...</span>
</code></pre>
<p>The C++ statements following the call to the JavaScript function will not be
executed. The exception will bubble up as a C++ exception of type <code>Napi::Error</code>,
until it is either caught while still in C++, or else automatically propagated as
a JavaScript exception when returning to JavaScript.</p>
<h3>Handling a N-API C++ exception</h3>
<pre class="hljs"><code>Napi::Function jsFunctionThatThrows = someObj.As&lt;Napi::Function&gt;();
Napi::Value result;
<span class="hljs-keyword">try</span> {
    result = jsFunctionThatThrows({ arg1, arg2 });
} <span class="hljs-keyword">catch</span> (<span class="hljs-keyword">const</span> Error&amp; e) {
    <span class="hljs-built_in">cerr</span> &lt;&lt; <span class="hljs-string">&quot;Caught JavaScript exception: &quot;</span> + e.what();
}
</code></pre>
<p>Since the exception was caught here, it will not be propagated as a JavaScript
exception.</p>
<p><a name="noexceptions"></a></p>
<h2>Handling Errors Without C++ Exceptions</h2>
<p>If C++ exceptions are disabled (for more info see: <a href="setup/">Setup</a>), then the
<code>Napi::Error</code> class does not extend <code>std::exception</code>. This means that any calls to
node-addon-api function do not throw a C++ exceptions. Instead, it raises
<em>pending</em> JavaScript exceptions and returns an <em>empty</em> <code>Napi::Value</code>.
The calling code should check <code>env.IsExceptionPending()</code> before attempting to use a
returned value, and may use methods on the <code>Napi::Env</code> class
to check for, get, and clear a pending JavaScript exception (for more info see: <a href="env/">Env</a>).
If the pending exception is not cleared, it will be thrown when the native code
returns to JavaScript.</p>
<h2>Examples with C++ exceptions disabled</h2>
<h3>Throwing a JS exception</h3>
<pre class="hljs"><code>Napi::Env env = ...
Napi::Error::New(env, <span class="hljs-string">&quot;Example exception&quot;</span>).ThrowAsJavaScriptException();
<span class="hljs-keyword">return</span>;
</code></pre>
<p>After throwing a JavaScript exception, the code should generally return
immediately from the native callback, after performing any necessary cleanup.</p>
<h3>Propagating a N-API JS exception</h3>
<pre class="hljs"><code>Napi::Env env = ...
Napi::Function jsFunctionThatThrows = someObj.As&lt;Napi::Function&gt;();
Napi::Value result = jsFunctionThatThrows({ arg1, arg2 });
<span class="hljs-keyword">if</span> (env.IsExceptionPending()) {
    Error e = env.GetAndClearPendingException();
    <span class="hljs-keyword">return</span> e.Value();
}
</code></pre>
<p>If env.IsExceptionPending() returns true a JavaScript exception is pending. To
let the exception propagate, the code should generally return immediately from
the native callback, after performing any necessary cleanup.</p>
<h3>Handling a N-API JS exception</h3>
<pre class="hljs"><code>Napi::Env env = ...
Napi::Function jsFunctionThatThrows = someObj.As&lt;Napi::Function&gt;();
Napi::Value result = jsFunctionThatThrows({ arg1, arg2 });
<span class="hljs-keyword">if</span> (env.IsExceptionPending()) {
    Napi::Error e = env.GetAndClearPendingException();
    <span class="hljs-built_in">cerr</span> &lt;&lt; <span class="hljs-string">&quot;Caught JavaScript exception: &quot;</span> + e.Message();
}
</code></pre>
<p>Since the exception was cleared here, it will not be propagated as a JavaScript
exception after the native callback returns.</p>
<h2>Calling N-API directly from a <strong>node-addon-api</strong> addon</h2>
<p><strong>node-addon-api</strong> provides macros for throwing errors in response to non-OK
<code>napi_status</code> results when calling <a href="https://nodejs.org/docs/latest/api/n-api.html" target="_blank" rel="noopener">N-API</a>
functions from within a native addon. These macros are defined differently
depending on whether C++ exceptions are enabled or not, but are available for
use in either case.</p>
<h3><code>NAPI_THROW(e, ...)</code></h3>
<p>This macro accepts a <code>Napi::Error</code>, throws it, and returns the value given as
the last parameter. If C++ exceptions are enabled (by defining
<code>NAPI_CPP_EXCEPTIONS</code> during the build), the return value will be ignored.</p>
<h3><code>NAPI_THROW_IF_FAILED(env, status, ...)</code></h3>
<p>This macro accepts a <code>Napi::Env</code> and a <code>napi_status</code>. It constructs an error
from the <code>napi_status</code>, throws it, and returns the value given as the last
parameter. If C++ exceptions are enabled (by defining <code>NAPI_CPP_EXCEPTIONS</code>
during the build), the return value will be ignored.</p>
<h3><code>NAPI_THROW_IF_FAILED_VOID(env, status)</code></h3>
<p>This macro accepts a <code>Napi::Env</code> and a <code>napi_status</code>. It constructs an error
from the <code>napi_status</code>, throws it, and returns.</p>
<h3><code>NAPI_FATAL_IF_FAILED(status, location, message)</code></h3>
<p>This macro accepts a <code>napi_status</code>, a C string indicating the location where the
error occurred, and a second C string for the message to display.</p>
</body></html></div>
        </main>
      </div>
    </div>
<footer class="app-footer text-md-left text-center">
  <div class="container">
    <div class="row align-items-center">
      <div class="col-md-2 mb-md-0 mb-4">
        <span class="pr-md-5 pr-0 py-3">
          <a href="https://runtimeverification.com" target="_blank">
            <picture>
              <source
                srcset="../../../../../assets/img/rv-logo-dark.png"
                media="(prefers-color-scheme: dark)"
              />
              <img
                class="pr-3 footer-logo"
                src="../../../../../assets/img/rv-logo.png"
                alt="Runtime Verification Inc logo"
              />
            </picture>
          </a>
        </span>
      </div>
      <div class="col-md-6 mb-md-0 mb-4"></div>
      <div class="col-md-4 text-md-right">
        <p class="copyright">
          &copy; 2020 Runtime Verification Inc. All right reserved.
        </p>
      </div>
    </div>
  </div>
</footer>

<script
  async
  src="https://www.googletagmanager.com/gtag/js?id=UA-163311512-1"
></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag() {
    dataLayer.push(arguments);
  }
  gtag("js", new Date());
  gtag("config", "UA-163311512-1");
</script>

    <script src="https://cdnjs.cloudflare.com/ajax/libs/Typist/1.2/typist.min.js"></script>
    <script src="../../../../../assets/js/index.js"></script>
  </body>
</html>
